File: frontend/RemoteDesktop/app/src/main/java/com/example/remotedesktop/Shared.kt
package com.example.remotedesktop

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
sealed interface ControlInput {
    @Serializable
    @SerialName("Cursor")
    data class Cursor(val x: Int, val y: Int, val relative: Boolean) :
        ControlInput
    @Serializable
    @SerialName("MouseButton")
    data class MouseButton(val detail: Int, val event: Int) : ControlInput
    @Serializable
    @SerialName("Key")
    data class Key(val detail: Int, val event: Int) : ControlInput
}

@Serializable
sealed interface Params {
    @Serializable
    @SerialName("Connect")
    data class Connect(val video_params: VideoParams, val session_info: SessionInfo) : Params
    @Serializable
    @SerialName("Stop")
    data class Stop(val session_info: SessionInfo) : Params
    @Serializable
    @SerialName("Query")
    object Query : Params
}

@Serializable
data class VideoParams(
    val bitrate: Int,
    val fps: Int,
    val width: Int,
    val height: Int
)

@Serializable
@Stable
data class SessionInfo(val id: String, val session: Session): Selectable {
    // This is a bad preview but it'll work for now
    override fun preview() = id
}

@Serializable
data class Entry(
    @SerialName("key")
    val keyRaw: Mut<String> = mutOf("key"),
    @SerialName("value")
    val valueRaw: Mut<String> = mutOf("value"),
) {
    var key : String by keyRaw
    var value : String by valueRaw
}

File: backend/remotedesktop/src/shared.rs
use std::{
    collections::hash_map::DefaultHasher,
    hash::{Hash, Hasher},
};

use serde::{Deserialize, Serialize};

#[derive(Deserialize)]
#[serde(tag = "type")]
pub enum ControlInput {
    Cursor { x: i32, y: i32, relative: bool },
    MouseButton { detail: u8, event: u8 },
    Key { detail: usize, event: u8 },
}

#[derive(Deserialize)]
#[serde(tag = "type")]
pub enum Params {
    Connect {
        video_params: VideoParams,
        session_info: SessionInfo,
    },
    Stop {
        session_info: SessionInfo,
    },
    Query,
}

#[derive(Deserialize)]
pub struct VideoParams {
    pub bitrate: u32,
    pub fps: u8,
    pub width: u32,
    pub height: u32,
}

#[derive(Deserialize, Serialize, PartialEq, Eq, Hash, Clone)]
pub struct SessionInfo {
    pub id: String,
    pub session: Session,
}

impl SessionInfo {
    pub fn display_name(&self) -> String {
        let mut hasher = DefaultHasher::new();
        self.hash(&mut hasher);
        let number = hasher.finish();
        format!(":{}", number);
        ":2".to_string()
    }
}

#[derive(Deserialize, Serialize, Hash, PartialEq, Eq, Clone)]
pub struct Session {
    pub name: String,
    pub desktop: Vec<String>,
    pub envs: Vec<Entry>,
}

#[derive(Deserialize, Serialize, Hash, PartialEq, Eq, Clone)]
pub struct Entry {
    pub key: String,
    pub value: String,
}

impl Entry {
    pub fn new(key: String, value: String) -> Entry {
        Entry { key, value }
    }
}

impl Entry {
    pub fn to_tuple(&self) -> (&str, &str) {
        (&self.key, &self.value)
    }
}
